crates/storage/nippy-jar/src/lib.rs

```diff
@@ -0,0 +1,407 @@
+            path: Some(path.to_path_buf()),
+        }
+    }
+
+    /// Adds [`compression::Zstd`] compression.
+    pub fn with_zstd(mut self, use_dict: bool, max_dict_size: usize) -> Self {
+        self.compressor =
+            Some(Compressors::Zstd(compression::Zstd::new(use_dict, max_dict_size, self.columns)));
+        self
+    }
+
+    /// Loads the file configuration and returns [`Self`].
+    pub fn load(path: &Path) -> Result<Self, NippyJarError> {
+        let mut file = File::open(path)?;
+        let mut obj: Self = bincode::deserialize_from(&mut file)?;
+
+        obj.path = Some(path.to_path_buf());
+        Ok(obj)
+    }
+
+    /// Returns the path from the data file
+    pub fn data_path(&self) -> PathBuf {
+        self.path.clone().expect("exists")
+    }
+
+    /// Returns the path from the index file
+    pub fn index_path(&self) -> PathBuf {
+        let data_path = self.data_path();
+        data_path
+            .parent()
+            .expect("exists")
+            .join(format!("{}.idx", data_path.file_name().expect("exists").to_string_lossy()))
+    }
+
+    /// If required, prepares any compression algorithm to an early pass of the data.
+    pub fn prepare(
+        &mut self,
+        columns: Vec<impl IntoIterator<Item = Vec<u8>>>,
+    ) -> Result<(), NippyJarError> {
+        // Makes any necessary preparations for the compressors
+        if let Some(compression) = &mut self.compressor {
+            compression.prepare_compression(columns)?;
+        }
+        Ok(())
+    }
+
+    /// Writes all necessary configuration to file.
+    pub fn freeze(
+        &mut self,
+        columns: Vec<impl IntoIterator<Item = Vec<u8>>>,
+        total_rows: u64,
+    ) -> Result<(), NippyJarError> {
+        let mut file = self.freeze_check(&columns)?;
+
+        self.freeze_config(&mut file)?;
+
+        // Special case for zstd that might use custom dictionaries/compressors per column
+        // If any other compression algorithm is added and uses a similar flow, then revisit
+        // implementation
+        let mut maybe_zstd_compressors = None;
+        if let Some(Compressors::Zstd(zstd)) = &self.compressor {
+            maybe_zstd_compressors = zstd.generate_compressors()?;
+        }
+
+        // Temporary buffer to avoid multiple reallocations if compressing to a buffer (eg. zstd w/
+        // dict)
+        let mut tmp_buf = Vec::with_capacity(100);
+
+        // Write all rows while taking all row start offsets
+        let mut row_number = 0u64;
+        let mut values_offsets = Vec::with_capacity(total_rows as usize * self.columns);
+        let mut column_iterators =
+            columns.into_iter().map(|v| v.into_iter()).collect::<Vec<_>>().into_iter();
+
+        loop {
+            let mut iterators = Vec::with_capacity(self.columns);
+
+            // Write the column value of each row
+            for (column_number, mut column_iter) in column_iterators.enumerate() {
```

Comment:
can we simplify the iteration logic here with
```suggestion
        let mut column_iterators =
            columns.into_iter().map(|v| v.into_iter()).collect::<Vec<_>>();

        for row_number in 0..=total_rows {
            // Write the column value of each row
            for (column_number, mut column_iter) in column_iterators.iter_mut().enumerate() {
```

so we use `for` instead of `loop` and reuse the column iterators that's created just once instead of collecting them into `iterators` over and over on each loop iteration?

Reply:
can't do it like that because of the `IntoIterator` interface. Which could be changed, but I'll wait until we start passing the table iterators and nail that down. 

i'll add a doc to revisit it another time.

Reply:
hmm, I think this actually works:
```diff
diff --git a/crates/storage/nippy-jar/src/lib.rs b/crates/storage/nippy-jar/src/lib.rs
index a4322f56e..4d50cae1c 100644
--- a/crates/storage/nippy-jar/src/lib.rs
+++ b/crates/storage/nippy-jar/src/lib.rs
@@ -108,16 +108,12 @@ impl NippyJar {
         let mut tmp_buf = Vec::with_capacity(100);
 
         // Write all rows while taking all row start offsets
-        let mut row_number = 0u64;
         let mut values_offsets = Vec::with_capacity(total_rows as usize * self.columns);
-        let mut column_iterators =
-            columns.into_iter().map(|v| v.into_iter()).collect::<Vec<_>>().into_iter();
-
-        loop {
-            let mut iterators = Vec::with_capacity(self.columns);
+        let mut column_iterators = columns.into_iter().map(|v| v.into_iter()).collect::<Vec<_>>();
 
+        for row_number in 0..total_rows {
             // Write the column value of each row
-            for (column_number, mut column_iter) in column_iterators.enumerate() {
+            for (column_number, column_iter) in column_iterators.iter_mut().enumerate() {
                 values_offsets.push(file.stream_position()? as usize);
 
                 match column_iter.next() {
@@ -147,16 +143,7 @@ impl NippyJar {
                         ))
                     }
                 }
-
-                iterators.push(column_iter);
             }
-
-            row_number += 1;
-            if row_number == total_rows {
-                break
-            }
-
-            column_iterators = iterators.into_iter();
         }
 
         // Write offset index to file
```